src/backend_model_instance.cc

```diff
@@ -547,36 +561,62 @@ TritonModelInstance::Initialize()
 Status
 TritonModelInstance::WarmUp()
 {
+  Status warmup_status = Status::Success;
   for (auto& sample : warmup_samples_) {
     LOG_VERBOSE(1) << "model '" << sample.requests_.back()->ModelName()
                    << "' instance " << Name() << " is running warmup sample '"
                    << sample.sample_name_ << "'";
 
-    std::promise<void> warmup_promise;
-    bool first_sample = true;
-
-    std::vector<TRITONBACKEND_Request*> triton_requests(1024);
-    triton_requests.clear();
-    for (auto& request : sample.requests_) {
-      request->SetReleaseCallback(
-          WarmupRequestComplete, first_sample ? &warmup_promise : nullptr);
-      first_sample = false;
+    std::vector<std::promise<void>> request_complete(sample.requests_.size());
+    std::vector<std::string> response_errors;
+    std::vector<std::pair<std::promise<void>, std::vector<std::string>*>>
+        response_complete(sample.requests_.size());
+
+    std::vector<TRITONBACKEND_Request*> triton_requests;
+    for (size_t i = 0; i < sample.requests_.size(); ++i) {
+      auto& request = sample.requests_[i];
+      request->SetReleaseCallback(WarmupRequestComplete, &request_complete[i]);
+      response_complete[i].second = &response_errors;
+      request->SetResponseCallback(
+          &warmup_allocator, nullptr, WarmupResponseComplete,
+          &response_complete[i]);
```

Comment:
Can you write a comment here summarizing this process? 

> ... set response callback for each warmup request and pass response_complete userp so we can check for errors from each async response and exit early if any warmup request fails ...

> ... use promises to signal when each response is complete async ... etc.

This part of the code is new to me so these kinds of summaries are very helpful when reading it later.

Otherwise, the WIP LGTM so far